### ISSUE TITLE
[FEATURE] Use Python 3.9's `Annotated[T, *metadata]` (PEP 593) to describe parameters

### ISSUE THREAD

**Issue by ewen-lbh** [tiangolo/typer#184]
### Is your feature request related to a problem

Not really, the current solution is already great.

### The solution you would like

Add an alternative syntax to describe optional parameters, [PEP 593](https://www.python.org/dev/peps/pep-0593/)'s `Annotated[T, *metadata]`. Now, strictly speaking of syntax, it's not much better, but it's built into python, and thus might help integration with IDEs in the future.

Help for optional parameters could be provided as the second argument of `Annotated`:

```python
def my_action(a: str, b: Annotated[bool, "description"], c: Annotated[str, "output"] = "~/.config/"):
```

is equivalent to

```python
def my_action(a: str, b: bool = Argument(..., help="description"), c: str = Option("~/.config/", help="output")):
```

As a bonus, there's no need to do a weird `thing: type = Argument(..., help="description")` when you want to specify help without giving a default value, just do `thing: Annotated[type, "description"]`

Also, it might be a little easier to implement since there's no messing with default values.

An `Annotated` parameter to the function will resolve to:

- `Argument` if it has no default value
- `Option` else

### Describe alternatives you've considered

Making it always resolve to `Argument` or always to `Option`, or use something like `Annotated[type, "description", form=Argument` for arguments and `Annotated[type, "description", form=Option]` for options. I'm not sure about this.

tbh this is debatable, but I figured I'd throw the idea out there, since this shiny new python 3.9 feature could be useful to Typer, as it seems that the use case described in the PEP is fitting:

> This PEP adds an Annotated type to the typing module to decorate existing types with context-specific metadata. Specifically, a type T can be annotated with metadata x via the typehint Annotated[T, x]. This metadata can be used for either static analysis or at runtime. If a library (or tool) encounters a typehint Annotated[T, x] and has no special logic for metadata x, it should ignore it and simply treat the type as T.

— [PEP 593, python.org](https://www.python.org/dev/peps/pep-0593/#motivation)

**Comment by baterflyrity** [comment]
It would be very useful. Cant wait implementation.

**Comment by mficek** [comment]
+1 for implementation. Can I help in any way? Is there any option to plug-in typer with this functionality?

**Comment by NowanIlfideme** [comment]
Similar PR in FastAPI repo hasn't been merged yet, it's possible this is waiting for Pydantic-Core to be released more broadly (I know Sebastian and Samuel Colvin talk :yum:) to fix this at an even more fundamental level.

**Comment by tiangolo** [comment]
Hello there! This was implemented in https://github.com/tiangolo/typer/pull/584, here are the new docs: https://typer.tiangolo.com/tutorial/arguments/optional/#an-alternative-cli-argument-declaration, it's available in Typer 0.9.0, just released :rocket:
